clang/test/CodeGenSYCL/regcall-cc-test.cpp

```diff
@@ -0,0 +1,360 @@
+
+// === TEST CASE: multi-level nested structs with only leaf fields of primitive
+// types. Unwrapping and merging should yield 2 32-bit integers
+
+struct B6 { int *a; int b; };
+struct C6 {
+  B6 b;
+  char x;
+  char y;
+
+  C6 foo() { return *this; }
+};
+
+// CHECK-DAG: %struct.C6 = type { %struct.B6, i8, i8 }
+// CHECK-DAG: %struct.B6 = type { i32 addrspace(4)*, i32 }
+
+template SYCL_DEVICE C6 __regcall func<C6>(C6 x);
+// CHECK-DAG: define weak_odr x86_regcallcc %struct.C6 @_Z16__regcall3__funcI2C6ET_S1_(%struct.B6 %{{[0-9a-zA-Z_.]+}}, i8 %{{[0-9a-zA-Z_.]+}}, i8 %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: a struct with sizeof() <= 2 is passed as a single 16-bit integer
+
+struct CharChar {
+  char a;
+  char b;
+
+  CharChar foo() { return *this; }
+};
+// CHECK-DAG: %struct.CharChar = type { i8, i8 }
+
+template SYCL_DEVICE CharChar __regcall func<CharChar>(CharChar x);
+// CHECK-DAG: define weak_odr x86_regcallcc i16 @_Z16__regcall3__funcI8CharCharET_S1_(i16 %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: a struct with sizeof() == 3-4 is passed as single 32-bit integer
+
+struct ShortShort {
+  short a;
+  short b;
+
+  ShortShort foo() { return *this; }
+};
+// CHECK-DAG: %struct.ShortShort = type { i16, i16 }
+
+template SYCL_DEVICE ShortShort __regcall func<ShortShort>(ShortShort x);
+// CHECK-DAG: define weak_odr x86_regcallcc i32 @_Z16__regcall3__funcI10ShortShortET_S1_(i32 %{{[0-9a-zA-Z_.]+}})
+
+struct CharShort {
+  char a;
+  short b;
+
+  CharShort foo() { return *this; }
+};
+// CHECK-DAG: %struct.CharShort = type { i8, i16 }
+
+template SYCL_DEVICE CharShort __regcall func<CharShort>(CharShort x);
+// CHECK-DAG: define weak_odr x86_regcallcc i32 @_Z16__regcall3__funcI9CharShortET_S1_(i32 %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: a struct with primitive single field element is just unwrapped
+
+struct Char {
+  char a;
+
+  Char foo() { return *this; }
+};
+// CHECK-DAG: %struct.Char = type { i8 }
+
+template SYCL_DEVICE Char __regcall func<Char>(Char x);
+// CHECK-DAG: define weak_odr x86_regcallcc i8 @_Z16__regcall3__funcI4CharET_S1_(i8 %{{[0-9a-zA-Z_.]+}})
+
+struct Float {
+  float a;
+
+  Float foo() { return *this; }
+};
+// CHECK-DAG: %struct.Float = type { float }
+
+template SYCL_DEVICE Float __regcall func<Float>(Float x);
+// CHECK-DAG: define weak_odr x86_regcallcc float @_Z16__regcall3__funcI5FloatET_S1_(float %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: a struct with sizeof() == 5-8 is passed as two 32-bit integers
+// 32-bit integers
+
+struct CharCharShortFloat {
+  char a, b;
+  short c;
+  float d;
+
+  CharCharShortFloat foo() { return *this; }
+};
+// CHECK-DAG: %struct.CharCharShortFloat = type { i8, i8, i16, float }
+
+template SYCL_DEVICE CharCharShortFloat __regcall func<CharCharShortFloat>(CharCharShortFloat x);
+// CHECK-DAG: define weak_odr x86_regcallcc [2 x i32] @_Z16__regcall3__funcI18CharCharShortFloatET_S1_([2 x i32] %{{[0-9a-zA-Z_.]+}})
```

Comment:
Could you please explain this?   The doc says floating point types are not merged.

Reply:
Thanks for catching! Will update the doc.